examples/seq2seq/seq2seq_trainer.py

```diff
@@ -0,0 +1,179 @@
+except ImportError:
+    from utils import label_smoothed_nll_loss
+
+
+logger = logging.getLogger(__name__)
+
+
+class Seq2SeqTrainer(Trainer):
+    def _get_train_sampler(self) -> Optional[torch.utils.data.sampler.Sampler]:
+        if isinstance(self.train_dataset, torch.utils.data.IterableDataset):
+            return None
+        elif is_torch_tpu_available():
+            return get_tpu_sampler(self.train_dataset)
+        else:
+            if self.args.sortish_sampler and self.args.n_gpu <= 1 and self.args.local_rank == -1:
+                return self.train_dataset.make_sortish_sampler(self.args.per_device_train_batch_size)
+
+            return (
+                RandomSampler(self.train_dataset)
+                if self.args.local_rank == -1
+                else DistributedSampler(self.train_dataset)
+            )
+
+    # override to support label smoothing
+    def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor, Any]]) -> float:
+        """
+        Perform a training step on a batch of inputs.
+
+        Subclass and override to inject custom behavior.
+
+        Args:
+            model (:obj:`nn.Module`):
+                The model to train.
+            inputs (:obj:`Dict[str, Union[torch.Tensor, Any]]`):
+                The inputs and targets of the model.
+
+                The dictionary will be unpacked before being fed to the model. Most models expect the targets under the
+                argument :obj:`labels`. Check your model's documentation for all accepted arguments.
+
+        Return:
+            :obj:`float`: The training loss on this batch.
+        """
+        model.train()
+        inputs = self._prepare_inputs(inputs)
+        labels = inputs.pop("labels")
+        if self.args.fp16 and _use_native_amp:
+            with autocast():
+                outputs = model(**inputs, use_cache=False)
+                logits = outputs[0]
+                loss = self._compute_loss(logits, labels, ignore_index=model.config.pad_token_id)
+        else:
+            outputs = model(**inputs, use_cache=False)
+            # We don't use .loss here since the model may return tuples instead of ModelOutput.
+            logits = outputs[0]
+            loss = self._compute_loss(logits, labels, ignore_index=model.config.pad_token_id)
+
+        if self.args.n_gpu > 1:
+            loss = loss.mean()  # mean() to average on multi-gpu parallel training
+
+        if self.args.gradient_accumulation_steps > 1:
+            loss = loss / self.args.gradient_accumulation_steps
+
+        if self.args.fp16 and _use_native_amp:
+            self.scaler.scale(loss).backward()
+        elif self.args.fp16 and _use_apex:
+            with amp.scale_loss(loss, self.optimizer) as scaled_loss:
+                scaled_loss.backward()
+        else:
+            loss.backward()
+
+        return loss.item()
+
+    def _compute_loss(self, logits, labels, ignore_index):
```

Comment:
@staticmethod ?

Reply:
any specific reason ?

Reply:
it allows the reader to quickly see that the method does not depend on `self`.
+ allows unittests to call `Seq2SeqTrainer.compute_loss` without instantiating an object/model